pkg/storage/mysql/mysql_test.go

```diff
@@ -43,51 +43,183 @@ func TestMySQLDatastoreAfterCloseIsNotReady(t *testing.T) {
+			firstTuple := tuple.NewTupleKey("doc:object_id_1", "relation", "user:user_1")
+			secondTuple := tuple.NewTupleKey("doc:object_id_2", "relation", "user:user_2")
+			thirdTuple := tuple.NewTupleKey("doc:object_id_3", "relation", "user:user_3")
+
+			err = sqlcommon.Write(ctx,
+				sqlcommon.NewDBInfo(ds.db, ds.stbl, sq.Expr("NOW()")),
+				store,
+				[]*openfgav1.TupleKeyWithoutCondition{},
+				[]*openfgav1.TupleKey{firstTuple},
+				time.Now())
+			require.NoError(t, err)
+
+			// Tweak time so that ULID is smaller.
+			err = sqlcommon.Write(ctx,
+				sqlcommon.NewDBInfo(ds.db, ds.stbl, sq.Expr("NOW()")),
+				store,
+				[]*openfgav1.TupleKeyWithoutCondition{},
+				[]*openfgav1.TupleKey{secondTuple},
+				time.Now().Add(time.Minute*-1))
+			require.NoError(t, err)
+
+			// Tweak time so that ULID is smaller.
+			err = sqlcommon.Write(ctx,
+				sqlcommon.NewDBInfo(ds.db, ds.stbl, sq.Expr("NOW()")),
+				store,
+				[]*openfgav1.TupleKeyWithoutCondition{},
+				[]*openfgav1.TupleKey{thirdTuple},
+				time.Now().Add(time.Minute*-2))
+			require.NoError(t, err)
+
+			iter, err := ds.Read(ctx, store, tuple.NewTupleKey("doc:", "relation", ""), storage.ReadOptions{})
+			defer iter.Stop()
+
+			require.NoError(t, err)
+
+			// We expect that objectID1 will return first because it is inserted first.
+			if tt.mixed {
+				curTuple, err := iter.Head(ctx)
+				require.NoError(t, err)
+				require.Equal(t, firstTuple, curTuple.GetKey())
+
+				// calling head should not change the order
+				curTuple, err = iter.Head(ctx)
+				require.NoError(t, err)
+				require.Equal(t, firstTuple, curTuple.GetKey())
+			}
+
+			curTuple, err := iter.Next(ctx)
+			require.NoError(t, err)
+			require.Equal(t, firstTuple, curTuple.GetKey())
+
+			curTuple, err = iter.Next(ctx)
+			require.NoError(t, err)
+			require.Equal(t, secondTuple, curTuple.GetKey())
+
+			if tt.mixed {
+				curTuple, err = iter.Head(ctx)
+				require.NoError(t, err)
+				require.Equal(t, thirdTuple, curTuple.GetKey())
+			}
+
+			curTuple, err = iter.Next(ctx)
+			require.NoError(t, err)
+			require.Equal(t, thirdTuple, curTuple.GetKey())
+
+			if tt.mixed {
+				_, err = iter.Head(ctx)
+				require.ErrorIs(t, err, storage.ErrIteratorDone)
+			}
+
+			_, err = iter.Next(ctx)
+			require.ErrorIs(t, err, storage.ErrIteratorDone)
+		})
+	}
+}
 
-	curTuple, err = iter.Next(ctx)
-	require.NoError(t, err)
-	require.Equal(t, secondTuple, curTuple.GetKey())
+func TestCtxCancel(t *testing.T) {
+	tests := []struct {
+		name  string
+		mixed bool
+	}{
+		{
+			name:  "nextOnly",
+			mixed: false,
+		},
+		{
+			name:  "mixed",
+			mixed: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testDatastore := storagefixtures.RunDatastoreTestContainer(t, "mysql")
+
+			uri := testDatastore.GetConnectionURI(true)
+			ds, err := New(uri, sqlcommon.NewConfig())
+			require.NoError(t, err)
+			defer ds.Close()
+
+			ctx, cancel := context.WithCancel(context.Background())
+
```

Comment:
```suggestion
			ctx, cancel := context.WithCancel(context.Background())
			defer cancel()
```

I know you have `cancel` somewhere below but this ensures that in execution doesn't reach that line, the cancel() still runs

Reply:
For this one, I actually want to invoke cancel at specific place (i.e., before calling Head and Next).  We are testing how context cancelled is being handled by the iterator.

Reply:
Yes, that is fine. I wasn't suggesting removing the `cancel()`, just adding the `defer cancel` as a catch-all